### ISSUE TITLE
[camera] Orientation is not sync with real devices' orientation while around multiple of 45 degrees

### ISSUE THREAD

**Issue by AlexV525** [flutter/flutter#78685]
TL;DR the orientation from the `CameraValue` is not synced with the device when the orientate angle reaches 45 degrees, the camera preview is rotating while the device's orientation hasn't changed yet.

https://user-images.githubusercontent.com/15884415/111867420-bc377d00-89ae-11eb-9bcd-8cf6f59891b4.mp4

With more advance usage, this issue will cause scale issue when these orientation are not sync. The following video is the example from [wechat_camera_picker](https://pub.dev/wechat_camera_picker).

https://user-images.githubusercontent.com/15884415/111867518-7202cb80-89af-11eb-8e3a-e977ffa37023.mp4

## Steps to Reproduce

1. Run `camera`'s example.
2. Enable device auto orientation.
3. Rotate the device around multiple of 45 degrees.

**Expected results:** The orientation should sync with the device.

**Actual results:** Preview is rotated around multiple of 45 degrees.

camera package version: 0.8.0
flutter version: 2.0.3

**Comment by AlexV525** [comment]
By searching "45" keyword in the plugin, shows a custom implementation with 45. https://github.com/flutter/plugins/search?q=45

cc @BeMacized

**Comment by darshankawar** [comment]
Verified using camera plugin's [example](https://pub.dev/packages/camera/example) and noticed the experience isn't same as that of other example mentioned above. 

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.1.0-13.0.pre.194, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.1.0-13.0.pre.194 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision e58ee0fbc7 (5 hours ago), 2021-03-22 13:06:06 +0800
    • Engine revision e6db3a275e
    • Dart version 2.13.0 (build 2.13.0-149.0.dev)

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 89.0.4389.90

! Doctor found issues in 1 category.

```

</details>

**Comment by AlexV525** [comment]
IMO the camera preview should not rotate according to the device's orientation. None of system's cameras will act like that, the real orientation that the photo/video has taken should be stored instead. So the important thing is make the camera preview not rotated.

After tried with some implementation, There are no way to match orientations manually by now, because: 
* Flutter use `Orientation` and `camera` use `DeviceOrientation`. `Orientation` only have `portrait` and `landscape`.
* The preview widget is rotating.

One more thing about the device orientation. Most of native apps take a rotated photo without unlock the system's orientation lock, and the orientation is still valid. Currently `camera` hasn't matched the behavior yet.

**Comment by BeMacized** [comment]
Took a quick look at this. The fact that the preview rotates to a different orientation than the current UI orientation, does indeed seem like unwanted behaviour. 

I do think we still need to keep rotating the preview so that it is still rotated correctly when the UI orientation changes, but relative to the UI orientation rather than the device orientation.
As you noticed `Orientation` doesn't allow users to differentiate between left vs right landscape mode, or regular portrait vs upside-down portrait, which makes this a bit more difficult. This was a major reason for me to add the custom device orientation implementation. 

I'd propose adding functionality to read UI orientation, similar to how checking device orientation is currently implemented, so the preview widget can use that instead.  

I did however see that a rather large refactor for the android side is happening in flutter/plugins#3651, which I'd prefer to land before making any changes of this sort. It's currently unclear to me how far off it is from being merged, so for now I'm going to hold off on building a fix.

**Comment by AlexV525** [comment]
Sure. Again the captured image/video's orientation may have to split out from the UI orientation, and not related to the device orientation's lock.

**Comment by marchelo** [comment]
I'm experiencing the same issue, my UI and camera orientation are not synchronized.

**Comment by Ahsanwarsi** [comment]
Hi everyone,
Was facing this issue for the last few days. By reading documentation & example project code carefully I found these functions very useful:
 **CameraController.lockCaptureOrientation**
 **CameraController.unlockCaptureOrientation**
 
 Locking the capture orientation did trick for me.
 
 @AlexV525 @marchelo

**Comment by AlexV525** [comment]
> **CameraController.lockCaptureOrientation**

😆 Haha. How could I take a landscape photo with orientation locked in portrait?

**Comment by BeMacized** [comment]
Several orientation related issues have been fixed in flutter/plugins#4059, including the preview rotating "too soon". 
I'll be closing this issue as I expect these fixes to have solved the problem.
If anyone is still running into issues regarding orientation using the new release, feel free to open a new issue, or post a message here so we may reopen this one.
